Objects/stringlib/find_max_char.h

```diff
@@ -20,23 +20,45 @@ Py_LOCAL_INLINE(Py_UCS4)
 STRINGLIB(find_max_char)(const STRINGLIB_CHAR *begin, const STRINGLIB_CHAR *end)
 {
     const unsigned char *p = (const unsigned char *) begin;
+    const unsigned char *_end = (const unsigned char *)end;
+    const size_t *aligned_end = (const size_t *)(_end - SIZEOF_SIZE_T);
+    const size_t *unrolled_end = aligned_end - 3;
+    unsigned char accumulator = 0;
+    /* Do not test each character individually, bit use bitwise OR and test
+       all characters at once. */
+    while (p < _end && !_Py_IS_ALIGNED(p, ALIGNOF_SIZE_T)) {
+        accumulator |= *p;
+        p += 1;
+    }
+    if (accumulator & 0x80) {
+        return 255;
+    } else if (p == end) {
+        return 127;
+    }
 
-    while (p < end) {
-        if (_Py_IS_ALIGNED(p, ALIGNOF_SIZE_T)) {
-            /* Help register allocation */
-            const unsigned char *_p = p;
-            while (_p + SIZEOF_SIZE_T <= end) {
-                size_t value = *(const size_t *) _p;
-                if (value & UCS1_ASCII_CHAR_MASK)
-                    return 255;
-                _p += SIZEOF_SIZE_T;
-            }
-            p = _p;
-            if (p == end)
-                break;
-        }
-        if (*p++ & 0x80)
+    /* On 64-bit platforms with 128-bit vectors (x86-64, arm64) the
+       compiler can load 4 size_t values into two 16-byte vectors and do a
+       vector bitwise OR. */
+    const size_t *_p = (const size_t *)p;
+    while (_p < unrolled_end) {
+        size_t value = _p[0] | _p[1] | _p[2] | _p[3];
+        if (value & UCS1_ASCII_CHAR_MASK) {
             return 255;
+        }
+        _p += 4;
+    }
+    size_t value = 0;
+    while (_p < aligned_end) {
+        value |= *_p;
+        _p += 1;
+    }
+    p = (const unsigned char *)_p;
+    while (p < _end) {
+        value |= *p;
+        p += 1;
+    }
```

Comment:
We could combine these two loops into a single loop right? Worst case we do a single loop over `32-1=31` bytes instead of two while loops (the loop over `_p`  (7 steps) and the loop over `p` (3 steps)).

Reply:
The more optimal solution is to do an unaligned load of a size_t at (end - SIZEOF_SIZE_T). That would remove the entire last while loop. I do not know if all architectures support unaligned loads however. I guess some could theoretically abort? 

I single loop over 31 bytes taking 31 steps is less optimal than 2 while loops taking 10 steps in total. (Or even 7 in the case of 64-bit size_t, that would be the majority of platforms).

Reply:
You can use the `memcpy` trick to do an unaligned load, it will be optimized by any reasonable compiler, for example:
```c
  size_t value;
  memcpy(&value, end - sizeof(value), sizeof(value));
```

Reply:
That's a very nice suggestion! That would simplify the code a lot.